FSI/Displacement.C

```diff
@@ -30,34 +43,45 @@ void preciceAdapter::FSI::Displacement::write(double * buffer, bool meshConnecti
         << exit(FatalError);
 }
 
+
 // return the displacement to use later in the velocity?
 void preciceAdapter::FSI::Displacement::read(double * buffer, const unsigned int dim)
 {
-    // For every element in the buffer
-    int bufferIndex = 0;
-
-    // For every boundary patch of the interface
-    for (uint j = 0; j < patchIDs_.size(); j++)
+    for (unsigned int j = 0; j < patchIDs_.size(); j++)
     {
-        int patchID = patchIDs_.at(j);
-
-        // Get the displacement on the patch
-        fixedValuePointPatchVectorField& pointDisplacementFluidPatch
-        (
-            refCast<fixedValuePointPatchVectorField>
-            (
-                pointDisplacement_->boundaryFieldRef()[patchID]
-            )
-        );
-
-        // For every cell of the patch
-        forAll(pointDisplacement_->boundaryField()[patchID], i)
-        {
-            // Set the displacement to the received one
-            pointDisplacementFluidPatch[i][0] = buffer[bufferIndex++];
-            pointDisplacementFluidPatch[i][1] = buffer[bufferIndex++];
-            if(dim ==3)
-                pointDisplacementFluidPatch[i][2] = buffer[bufferIndex++];
+        // Get the ID of the current patch
+        const unsigned int patchID = patchIDs_.at(j);
+
+        if (this->locationsType_ == "faceCenters" || this->locationsType_ == "faceCentres") {
+
+          // the boundaryCellDisplacement is a vector and ordered according to the iterator j
+          // and not according to the patchID
+          // First, copy the buffer data into the center based vectorFields on each interface patch
+          for (int i = 0; i < cellDisplacement_->boundaryField()[patchID].size(); ++i) {
+            for (unsigned int d = 0; d < dim; ++d)
+              cellDisplacement_->boundaryFieldRef()[patchID][i][d] = buffer[i * dim + d];
+          }
+          // Get a reference to the displacement on the point patch in order to overwrite it
+          vectorField &pointDisplacementFluidPatch(
+              refCast<vectorField>(
+                  pointDisplacement_->boundaryFieldRef()[patchID]));
+
+          // Overwrite the node based patch using the interpolation objects and the cell based vector field
+          // Afterwards, continue as usual
+          pointDisplacementFluidPatch = interpolationObjects_[j]->faceToPointInterpolate(cellDisplacement_->boundaryField()[patchID]);
+
+        } else if (this->locationsType_ == "faceNodes") {
+
+          // Get the displacement on the patch
+          fixedValuePointPatchVectorField &pointDisplacementFluidPatch(
+              refCast<fixedValuePointPatchVectorField>(
+                  pointDisplacement_->boundaryFieldRef()[patchID]));
+
+          // Overwrite the nodes on the interface directly
+          for (int i = 0; i < pointDisplacement_->boundaryFieldRef()[patchID].size(); ++i) {
```

Comment:
Why not using the same `forAll` construct here?

Reply:
I don't like it that much. I replaced it now by `forAll`